오픈마켓 [STEP 1] 토니, 앨리

Allie-e opened the pull request:
찌루루 @jae57 안녕하세요! 😆

토니 앨리입니다ㅎㅎ

STEP 1 구현완료하여 PR 드립니다.

URLSession을 처음 사용해봐서 개념을 이해하고 적용하는데 시간이 좀 걸렸습니다.

아직 확실하게 이해하고 있는 게 아니다보니 질문드리기가 쉽지 않네요..😥

부족한 점이 많겠지만 리뷰 잘 부탁드립니다 🙏🏻

## STEP 1 구현 내용


### Model

- 상품의 정보와 상품 페이지 정보를 담는 모델 타입인 ProductList, ProductInformation, ProductImage, Vendors를 구현하여 Codable 프로토콜 채택
- HTTPMethod 열거형을 만들어 namespace로 활용
- JSON data를 원하는 형식으로 decoding하기 위한 Parser열거형 내부 `static func decode()`구현
- 서버와 데이터 통신을 하기 위해 APIManager 타입 구현

## 고민했던 점


### URLSession을 싱글톤으로 사용하는 이유?

저희가 URLSession을 공부하다가, shared 라는 키워드를 보고 싱글톤패턴인 것을 알아차렸습니다. 그래서 왜 싱글톤이여야 하나라는 생각으로 찾아보고, 또 고민을 많이 해보았지만 명쾌한 결론을 얻지 못해서 질문 드립니다. 😵‍💫 

우선 싱글톤은 하나의 객체만을 생성하며, 그 하나의 객체로 여러곳에서 사용할 수 있도록 한다고 알고있습니다. 다시한번 풀어해석하면, 하나의 인스턴스를 메모리에 등록해서 여러스레드가 동시에 해당 인스턴스를 공유하여 사용하는 것이 싱글톤이라고 알게 되었는데, 장점은 1. 메모리를 단 한번만 사용. 2. 객체 접근시간이 줄어들게 됨 3. 전역범위에서 상태, 데이터 전달이 쉬워짐. 이라고 배웠습니다. 하지만, 저희가 URLSession 자체가 많이 익숙하지않아 왜 싱글톤인지 깨닫지 못했는데, 찌루루의 가르침 부탁드립니다. 🥺

### **HealthChecker의 필요성**

API 문서 내 healthChecker를 통해 네트워킹이 성공했는지를 확인하는 것 같습니다. 정상적으로 네트워킹이 되었다면, “OK”를 출력하도록 하고있는데, 필요성에 대해 의문이 들었습니다. 이러한 절차가 꼭 필요한 것인지 궁금합니다.🧐

## 조언을 얻고 싶은 부분

### completionHandler의 매개변수 error의 역할

URLSession의 dataTask에서 completionHandler의 매개변수로 포함되어 있는 `error`의 용도에 대해 의문점이 생겼습니다. response시 오류가 있으면 statusCode로 알수있는데 error가 왜 있는지 궁금합니다! 추가적으로 서버 네트워킹이 성공하거나 실패했을 때에도 항상 `nil`인 것을 확인했는데 어떤 경우에 error라는 매개변수를 사용하는지도 궁금합니다 🤔

```swift
let task = URLSession.shared.dataTask(with: request) { data, response, error in
    guard error == nil else {
        return
    }
            
    guard let statusCode = (response as? HTTPURLResponse)?.statusCode else {
        return
    }
            
    guard self.successRange.contains(statusCode) else {
        return
    }
            
    guard let data = data else {
        print(String(describing: error))
        self.semaphore.signal()
                
        return
    }
```

Allie-e committed:
feat: ProductInformation 모델 타입 구현 및 파일분리 #1 #2

oksunwoo committed:
feat: ProductList 타입 구현 #1

Allie-e committed:
feat: JSONData decode를 위한 타입메소드 구현 #1

oksunwoo committed:
test: Json 데이터 디코딩 테스트 #3

Allie-e committed:
feat: APIManager 타입 구현 및 requestHealthChecker() 추가 #1

oksunwoo committed:
feat: APIManager타입 내부에 requestProductInformation 메서드 구현 #1

- Parser에 Generic 적용 및 내부 구현 수정
- APIManager 타입 내부에 product 정보를 저장할 product 추가

Allie-e committed:
feat: APIManager 타입에 requestProductList() 메소드 추가 #1

oksunwoo committed:
test: Parser 타입메서드의 변경으로 인한 테스트수정 #3

Allie-e committed:
refactor: URLRequest httpMethod의 하드코딩 개선 #2 #4

oksunwoo committed:
refactor: dataTask 내부에 error와 response를 처리 #4

Allie-e committed:
refactor: dataTask() 메서드 내부 error, response 처리 #4

oksunwoo committed:
refactor: HTTPMethod case를 타입프로퍼티로 변경 #4

Allie-e committed:
feat: ProductImage 타입 구현 #1

oksunwoo committed:
feat: Vendors 타입 구현 #1

Allie-e committed:
refactor: ProductInformation 타입에 images, vendors 프로퍼티 추가 #4

oksunwoo committed:
test: APIManager test코드 구현 #3

Allie-e committed:
test: Mock테스트를 위한 URLSessionProtocol 구현 #3

oksunwoo committed:
chore: 불필요한 파일 제거 #4

jae57 commented on OpenMarket/ProductsTests/ProductsTests.swift:
```diff
@@ -0,0 +1,25 @@
+    }
+    
+    func test_decoding된_JSON데이터의_productList의_pageNumber가_1인지() {
+        let decoder = try! Parser<ProductList>.decode(from: sut!)
```
decoder는 디코드를 하는 주체라고 생각될 것 같아서 좀 더 디코딩된 결과를 의미하는 변수명이면 더 좋을것 같아요!

Allie-e replied:
넵 말씀해주신 것 처럼 decoder라는 네이밍은 다소 모호한 것 같아서 decode의 결과라는 의미를 담아 `decodeResult`로 수정해주었습니다 😊

jae57 commented on OpenMarket/OpenMarket/Model/APIManager.swift:
```diff
@@ -0,0 +1,118 @@
+            
+            guard self.successRange.contains(statusCode) else {
+                return
+            }
```
이 세가지 조건이 모두 별다른 동작없이 return 시키는 거라면 하나의 guard문으로 써줘도 될 것 같아요!
```swift
guard error == nil,
      let statusCode = (response as? HTTPURLResponse)?.statusCode,
      self.successRange.contains(statusCode) else {
       return
}
```

oksunwoo replied:
오 저희도 그 의견에 동의를 해서 코드를 제안해주신 것과 같이 변경했습니다 😇

Allie-e replied:
추가적으로, APIManager 타입의 메소드 내부 로직을 변경하면서, 세가지 조건에 대해 error 처리를 하도록 수정해주었습니다!

jae57 commented on OpenMarket/OpenMarket/Model/APIManager.swift:
```diff
@@ -0,0 +1,118 @@
+            
+            guard let data = data else {
+                print(String(describing: error))
+                self.semaphore.signal()
```
위 세가지 조건에서는 별다른 동작없이 return 하는데 여기서만 error를 print하고 semaphore.signal()을 주는 이유가 따로 있을까요?

oksunwoo replied:
고민을 해본 결과, 각각의 상황에서 에러 메시지를 출력해주는 편이 조금 더 나을 것이라고 판단해서, 각각 에러 처리를 해주었습니다.

```swift
guard error == nil else {
                return print(URLSessionError.requestFail.errorDescription)
            }
            
guard let statusCode = (response as? HTTPURLResponse)?.statusCode,
              self.successRange.contains(statusCode) else {
                return print(URLSessionError.statusCodeError.errorDescription)
            }
            
guard let data = data else {
             print(URLSessionError.invalidData.errorDescription)
             self.semaphore.signal()
                
                return
            }
```

semaphore signal은 에러가 발생한 경우, 프로그램이 wait상태에서 기다릴까봐 에러구현부에도 넣어주었습니다. 😆

Allie-e replied:
추가적으로 이부분 escaping closure로 수정해주었습니다~!

jae57 commented on OpenMarket/OpenMarket/Model/APIManager.swift:
```diff
@@ -0,0 +1,118 @@
+        }
+        
+        task.resume()
+        self.semaphore.wait()
```
해당 semaphore를 쓰시게 된 이유가 무엇인지 궁금합니다!
태스크 내에서 productList 값을 이용해 어떤 처리를 하는 부분이 없고, 
단순히 productList에 새로운 값을 업데이트 해주는 것이다보니
경쟁상태에 놓이지 않을 것 같은데, 이것 외에 예상하신 것이 있을까요?

oksunwoo replied:
우선 저희가 일반적인 클로저를 쓰다보니, 클로저 내부에서 값이 return이 될 때 아무것도 담기지 않는 현상을 발견할 수 있었습니다. 그래서 고민끝에 `semaphore`를 써주게 되었는데, 말씀해주신 경쟁상태보다는 `signal()`과 `wait()`을 통해 쓰레드를 멈춘상태에서 값을 받아오려고 했던 의도가 있었습니다. `semaphore`의 사용목적에 맞지 않는 것 같다는  상의를 오래도록 한 끝에, `semaphore` 대신 `escaping closure`를 사용하는 방법으로 수정했습니다. `escaping closure`를 사용하여, 함수가 종료된 이후에도 `closure`가 실행되어 data를 성공적으로 받아올 수 있었습니다.  

```swift
func requestProductInformation(productID: Int, completionHandler: @escaping (Result<ProductInformation, Error>) -> Void) {
        guard let url = URL(string: apiHost + "/api/products/" + "\(productID)") else {
            completionHandler(.failure(URLSessionError.urlIsNil))
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = HTTPMethod.get
        createDataTask(with: request, completionHandler)
    }
```

jae57 commented on OpenMarket/OpenMarket/Model/APIManager.swift:
```diff
@@ -0,0 +1,118 @@
+    
+    func requestProductList() {
+        let url = apiHost + "/api/products?page-no=1&items-per-page=10"
+        var request = URLRequest(url: URL(string: url)!,timeoutInterval: Double.infinity)
```
강제언래핑(!) 보다는 좀 더 안전한 방법으로 꺼내볼까요?

Allie-e replied:
넵 강제언래핑 지양하기위해 URL 안전하게 옵셔널 바인딩 해주었습니다 😊

jae57 commented on OpenMarket/OpenMarket/Model/APIManager.swift:
```diff
@@ -0,0 +1,118 @@
+        }
+        
+        task.resume()
+        self.semaphore.wait()
```
이러한 비슷한 동작을 하는 코드가 아래 `requestProductInformation`, `requestProductList` 에서 반복되고 있는데, 공통된 로직을 뽑아내보면 어떨까요?

Allie-e replied:
두 함수에서 코드가 반복되는 것을 줄이기 위해 `createDataTask() `메소드를 구현하여 각 함수에서 호출해서 재사용 할 수 있게 수정해주었습니다~!

jae57 commented on OpenMarket/OpenMarket/Model/APIManager.swift:
```diff
@@ -0,0 +1,118 @@
+    }
+    
+    func requestProductList() {
+        let url = apiHost + "/api/products?page-no=1&items-per-page=10"
```
`page-no`, `items-per-page` 와 같은 쿼리는 현재 값이 고정되어있는데 메서드에 파라미터를 두어 좀 더 활용성을 높혀보는 건 어떨까요? (requestProductInformation 메서드에서 productID를 받아서 url을 구성한 것 처럼요!)

Allie-e replied:
해당 함수의 파라미터로 `page_no`, `items_per_page`를 받아와서 `URLComponents`를 활용해서 `queryItem`으로 append 해주는 방식으로 수정해주었습니다ㅎㅎ

jae57 commented on OpenMarket/OpenMarket/Model/APIManager.swift:
```diff
@@ -0,0 +1,118 @@
+                return
+            }
+            
+            self.healthChecker = String(data: data, encoding: .utf8)!
```
강제 언래핑(!)을 좀 더 안전하게 처리해보면 어떨까요?

Allie-e replied:
해당 함수 내부 로직을 수정하면서 강제언래핑 삭제해주었습니다 ✔️

jae57 commented on OpenMarket/OpenMarket/Model/APIManager.swift:
```diff
@@ -0,0 +1,118 @@
+    var healthChecker: String?
+    var product: ProductInformation?
+    var productList: ProductList?
+    var semaphore = DispatchSemaphore (value: 0)
```
여기는 괄호직전에 띄어쓰기가 있네요? 👀

Allie-e replied:
앗 띄어쓰기 수정해주었습니다!

jae57 commented on OpenMarket/OpenMarket/Model/APIManager.swift:
```diff
@@ -0,0 +1,118 @@
+    let apiHost = "https://market-training.yagom-academy.kr/"
+    var healthChecker: String?
+    var product: ProductInformation?
+    var productList: ProductList?
```
APIManager의 역할은 무엇인가요? 
현재는 API통신 뿐 아니라 통신의 결과를 저장하는 역할도 하고 있네요! 
꼭 필요한 역할 하나만 해줄 수 있도록 바꿔보면 어떨까요?

oksunwoo replied:
기존에 있던 `APIManager`의 역할은 API통신도 하고 결과도 저장하는 역할을 해서 하나의 클래스가 너무 많은 역할을 하고 있다고 느껴서, 구현 방식을 `completionHandler`를 통해서 추후에 data를 처리할 수 있도록 변경하였습니다. 그래서 현재 `APIManager`는 API 통신의 역할만 하게 되었습니다 😙

jae57 reviewed (commented):
@oksunwoo, @wooyani77 
안녕하세요, 토니, 앨리! 
고생하셨습니다!!

## 고민했던 점
> URLSession을 싱글톤으로 사용하는 이유?

네트워크 통신을 간편히 할 수 있도록 만들어둔 객체이므로 어디서든 접근하여 사용할 수 있어야하고, 통신은 아무래도 자주 쓰이다 보니 생성에 쓰이는 비용을 줄여보려 싱글턴을 사용하게 되었지 않나 싶네요!

> HealthChecker의 필요성

서버의 상태를 간편하게 확인할 수 있겠어요!
백엔드나 프론트엔드나 에러를 디버깅 할 때 일단 서버는 정상적으로 띄워져 있는지 간편하게 확인하는 용도로도 쓸 수 있을거 같아요! 좀 더 앱 로직 내에 활용해본다고 하면, 통신을 하기 전에 해당 API를 통해 서버의 상태를 확인하고 비정상적이라면 통신을 하지 않고 얼럿을 띄워두고, 서버가 정상적으로 돌아왔을 때 통신을 하도록하면 앱이 좀 더 효율적으로 동작할 수 있을 거 같습니다. 

## 조언을 얻고 싶은 부분 
> 어떤 경우에 error라는 매개변수를 사용하는지도 궁금합니다

statusCode는 서버상에서 응답을 보낼 때 지정해준 숫자 코드가 담겨서 오는 부분이고, 
(따라서 일단 통신은 잘 되었다는 뜻이겠죠! )
error 부분은 통신 중에 발생한 에러들이 들어옵니다. 
아주 간단한 예시로, 와이파이 연결을 끊으면 요청자체가 서버에 도달하지 못하고, fail됩니다
이때 error로 들어오는 것이죠! `The Internet connection appears to be offline.` 이라는 description을 가진 error가 발생하네요!

![image](https://user-images.githubusercontent.com/15043840/148653776-47b42c56-ae0d-4be0-b07f-47fb4edf5475.png)
https://developer.apple.com/documentation/foundation/urlsession/1407613-datatask
https://developer.apple.com/forums/thread/114119

Allie-e committed:
test: test 코드 네이밍 수정 #3

oksunwoo committed:
refactor: 메서드 내부의 guard문을 하나의 guard문으로 수정 #4

Allie-e committed:
refactor: request에 따른 에러를 처리하도록 error타입 구현 및 적용 #4

oksunwoo committed:
refactor: APIManager 타입 내부 메서드의 클로저를 escaping 클로저로 변경 #4

- 중복코드를 피하기위해 createDataTask메서드를 재사용
- parser 타입 내부에 에러처리코드 추가
- ParserError 에러 타입 구현

Allie-e committed:
refactor: requestHealthChecker 내부 로직 수정 #4

Allie-e committed:
refactor: URL 강제 옵셔널 바인딩 수정 #4

oksunwoo committed:
refactor: URL구성방식을 메서드에 파라미터를 활용하는 방법으로 변경 #4

- URLSessionError에 url관련 Error 케이스 추가
- url 바인딩 실패시 completionHandler를 이용하여 에러처리하도록 변경

Allie-e committed:
refactor: APIManager 내부의 사용되지 않는 프로퍼티 삭제 #4

Allie-e committed:
test: APIManager 내부 메소드 로직 변경에따른 테스트코드 수정 #3

oksunwoo committed:
test: 불필요한 try 키워드 제거 #3

jae57 commented:
@oksunwoo, @wooyani77
STEP1 필수 구현사항 모두 잘 구현해주셨고, 남긴 코멘트 모두 반영해주셨네요!!
고생하셨습니다! 😀👏
